### PR TITLE
cluster: fix failure handling code logic

### PR DESCRIPTION
**Pull request by xiang90**
The previous code logic is unclear and error-prone since the if
condidtion describes the intended behavior not the state condition.

Changing the condition to state condition, and handling the condition
correctly.

### CODE REVIEW

File: pkg/cluster/cluster.go [hunk 1]
```diff
@@ -235,9 +235,13 @@ func (c *Cluster) run(stopC <-chan struct{}, wg *sync.WaitGroup) {
 				// TODO: we can't handle another upgrade while an upgrade is in progress
 				c.logger.Infof("spec update: from: %v to: %v", c.cluster.Spec, event.cluster.Spec)
 				c.cluster = event.cluster
+
 			case eventDeleteCluster:
+				c.status.SetReason("cluster is deleted by the user")
```
Comment (hongchaodeng) [inline pkg/cluster/cluster.go]: TPR is already deleted. Setting status isn't helpful.

File: pkg/cluster/cluster.go [hunk 2]
```diff
@@ -235,9 +235,13 @@ func (c *Cluster) run(stopC <-chan struct{}, wg *sync.WaitGroup) {
...
+				clusterFailed = true
```
Comment (hongchaodeng) [inline pkg/cluster/cluster.go]: We probably need another state called "clusterDeleted" and wouldn't need to report failed on that.

File: pkg/cluster/cluster.go (hunk 1 above)
Comment (xiang90) [inline pkg/cluster/cluster.go]: this is just to make code clear. probably i should change this to a logging message?

File: pkg/cluster/cluster.go (hunk 2 above)
Comment (xiang90) [inline pkg/cluster/cluster.go]: we can do that in another pr.

File: pkg/cluster/cluster.go (hunk 1 above)
Comment (hongchaodeng) [inline pkg/cluster/cluster.go]: There is another logging message
```
"deleting the failed cluster"
```

We'd better have one message for either "failed deleting" and "normal deleting". But definitely not two together.

**Comment by xiang90** [comment]
@hongchaodeng fixed. PTAL.

**Comment by hongchaodeng** [comment]
> We'd better have one message for either "failed deleting" and "normal deleting". But definitely not two together.

I probably want this eventually. But I'm fine with this PR. So LGTM here

File: pkg/cluster/cluster.go [hunk 3]
```diff
@@ -205,17 +205,18 @@ func (c *Cluster) send(ev *clusterEvent) {
...
 
 func (c *Cluster) run(stopC <-chan struct{}, wg *sync.WaitGroup) {
-	needDeleteCluster := true
+	clusterFailed := false
 
 	defer func() {
-		if needDeleteCluster {
-			c.logger.Infof("deleting cluster")
+		if clusterFailed {
+			c.reportFailedStatus()
```
Comment (hongchaodeng) [inline pkg/cluster/cluster.go]: Make this async?

This will block controller unboundedly

Comment (xiang90) [inline pkg/cluster/cluster.go]: why cluster will block the controller? i do not think it ever should. if cluster blocks controller, then there is something wrong somewhere else i guess.

Comment (hongchaodeng) [inline pkg/cluster/cluster.go]: reportFailedStatus() is unbounded.

On restore, controller will wait all cluster run() exit, using the passed "wg".

Comment (xiang90) [inline pkg/cluster/cluster.go]: > On restore, controller will wait all cluster run() exit, using the passed "wg".

this is wrong, and we should fix it. controller should never wait for a specific cluster.

Comment (xiang90) [inline pkg/cluster/cluster.go]: also i am not super concerned about this case right now. if reporting fails, then the cluster cannot connect to API server. Controller cannot do anything meaningful either. But we should fix the root problem: a cluster event should never blocks the controller.

Comment (hongchaodeng) [inline pkg/cluster/cluster.go]: > a cluster event should never blocks the controller.

This is not only cluster event, but also controller restore case. Waiting all exits only happens on controller restore.

> if reporting fails, then the cluster cannot connect to API server.

Add a comment?

**Comment by hongchaodeng** [comment]
@etcd-bot retest this please

**Comment by hongchaodeng** [comment]
@etcd-bot retest this please

https://github.com/coreos/etcd-operator/issues/627
